File: src/main/java/com/melodical/backend/entity/Comment.java
package com.melodical.backend.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "comments", indexes = {
    @Index(name = "idx_musical_id", columnList = "musical_id"),
    @Index(name = "idx_user_id", columnList = "user_id"),
    @Index(name = "idx_parent_id", columnList = "parent_id"),
    @Index(name = "idx_created_at", columnList = "created_at")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Comment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "musical_id", nullable = false)
    private Musical musical;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String content;

    // 대댓글 관계: 부모 댓글 (null이면 최상위 댓글)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Comment parent;

    // 자식 댓글 목록 (대댓글들)
    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Comment> replies = new ArrayList<>();

    // 대댓글 깊이 (0: 최상위, 1: 1단계 대댓글, 2: 2단계 대댓글 등)
    @Column(name = "depth", nullable = false)
    @Builder.Default
    private Integer depth = 0;

    // 삭제 여부 (실제 삭제 대신 소프트 삭제)
    @Column(name = "is_deleted", nullable = false)
    @Builder.Default
    private Boolean isDeleted = false;

    // 좋아요 수 (캐시용)
    @Column(name = "like_count", nullable = false)
    @Builder.Default
    private Integer likeCount = 0;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    public void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
        if (depth == null) {
            depth = parent != null ? parent.getDepth() + 1 : 0;
        }
        if (isDeleted == null) {
            isDeleted = false;
        }
        if (likeCount == null) {
            likeCount = 0;
        }
    }

    @PreUpdate
    public void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // 대댓글 추가 편의 메서드
    public void addReply(Comment reply) {
        replies.add(reply);
        reply.setParent(this);
        reply.setDepth(this.depth + 1);
    }

    // 삭제 처리 (소프트 삭제)
    public void softDelete() {
        this.isDeleted = true;
        this.content = "삭제된 댓글입니다.";
        this.updatedAt = LocalDateTime.now();
    }
}

File: lib/models/comment.dart
class Comment {
  final int id;
  final int musicalId;
  final String? musicalTitle;  // 뮤지컬 제목 추가
  final int userId;
  final String username;
  final String content;
  final DateTime createdAt;
  final DateTime? updatedAt;
  
  // 대댓글 관련 필드
  final int? parentId;
  final int depth;
  final bool isDeleted;
  final int likeCount;
  final List<Comment> replies;

  Comment({
    required this.id,
    required this.musicalId,
    this.musicalTitle,
    required this.userId,
    required this.username,
    required this.content,
    required this.createdAt,
    this.updatedAt,
    this.parentId,
    this.depth = 0,
    this.isDeleted = false,
    this.likeCount = 0,
    this.replies = const [],
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['id'],
      musicalId: json['musicalId'],
      musicalTitle: json['musicalTitle'],
      userId: json['userId'],
      username: json['username'] ?? 'Unknown',
      content: json['content'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      parentId: json['parentId'],
      depth: json['depth'] ?? 0,
      isDeleted: json['isDeleted'] ?? false,
      likeCount: json['likeCount'] ?? 0,
      replies: json['replies'] != null
          ? (json['replies'] as List).map((r) => Comment.fromJson(r)).toList()
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'musicalId': musicalId,
      'musicalTitle': musicalTitle,
      'userId': userId,
      'username': username,
      'content': content,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'parentId': parentId,
      'depth': depth,
      'isDeleted': isDeleted,
      'likeCount': likeCount,
      'replies': replies.map((r) => r.toJson()).toList(),
    };
  }

  // 좋아요 개수 업데이트를 위한 copyWith 메서드
  Comment copyWith({
    int? id,
    int? musicalId,
    String? musicalTitle,
    int? userId,
    String? username,
    String? content,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? parentId,
    int? depth,
    bool? isDeleted,
    int? likeCount,
    List<Comment>? replies,
  }) {
    return Comment(
      id: id ?? this.id,
      musicalId: musicalId ?? this.musicalId,
      musicalTitle: musicalTitle ?? this.musicalTitle,
      userId: userId ?? this.userId,
      username: username ?? this.username,
      content: content ?? this.content,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      parentId: parentId ?? this.parentId,
      depth: depth ?? this.depth,
      isDeleted: isDeleted ?? this.isDeleted,
      likeCount: likeCount ?? this.likeCount,
      replies: replies ?? this.replies,
    );
  }
}

class CommentRequest {
  final int musicalId;
  final String content;
  final int? parentId; // 대댓글인 경우 부모 댓글 ID

  CommentRequest({
    required this.musicalId,
    required this.content,
    this.parentId,
  });

  Map<String, dynamic> toJson() {
    return {
      'musicalId': musicalId,
      'content': content,
      if (parentId != null) 'parentId': parentId,
    };
  }
}
